# SPRINT 03 TASCA M3 T02

# 1.- Exercici 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

->Airlines Delay: Airline on-time statistics and delay causes

Important: Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.

In [1]:
# Cargar la tabla:
import pandas as pd

Delay_Table = pd.read_csv('DelayedFlights.csv')
Delay_Table[:10]

# Veo al final del resumen que la tabla está formada por 30 columnas

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   
5           6  2008      1           3          4   1937.0        1830   
6          10  2008      1           3          4    706.0         700   
7          11  2008      1           3          4   1644.0        1510   
8          15  2008      1           3          4   1029.0        1020   
9          16  2008      1           3          4   1452.0        1425   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   
5   2037.0        1940            WN  ...     3.0     7.0          0   
6    916.0         915            WN  ...     5.0    19.0          0   
7   1845.0        1725            WN  ...     6.0     8.0          0   
8   1021.0        1010            WN  ...     6.0     9.0          0   
9   1640.0        1625            WN  ...     7.0     8.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   
5                 N         0          10.0           0.0      0.0   
6                 N         0           NaN           NaN      NaN   
7                 N         0           8.0           0.0      0.0   
8                 N         0           NaN           NaN      NaN   
9                 N         0           3.0           0.0      0.0   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  
5           0.0               47.0  
6           NaN                NaN  
7           0.0               72.0  
8           NaN                NaN  
9           0.0               12.0  

[10 rows x 30 columns]

### Conocer el número de columnas:

In [2]:
num_columnas = Delay_Table.shape[1]
print ('la tabla Delay_Table está compuesta por:', num_columnas,'columnas')

la tabla Delay_Table está compuesta por: 30 columnas


### Conocer el número de filas:

In [3]:
num_filas = Delay_Table.shape[0]
print ('la tabla Delay_Table está compuesta por:', num_filas,'filas')

la tabla Delay_Table está compuesta por: 1936758 filas


### Listar los nombres de todas las columnas:

In [4]:
nombres_columnas = Delay_Table.columns
nombres_columnas

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

### Conocer y entener mejor los contenidos de las columnas,

In [5]:
print('Los nombres de las 30 columnas son:')
for col in Delay_Table.columns:
    print(col)

Los nombres de las 30 columnas son:
Unnamed: 0
Year
Month
DayofMonth
DayOfWeek
DepTime
CRSDepTime
ArrTime
CRSArrTime
UniqueCarrier
FlightNum
TailNum
ActualElapsedTime
CRSElapsedTime
AirTime
ArrDelay
DepDelay
Origin
Dest
Distance
TaxiIn
TaxiOut
Cancelled
CancellationCode
Diverted
CarrierDelay
WeatherDelay
NASDelay
SecurityDelay
LateAircraftDelay


### Ver el tipo de dato que tiene cada columna:

In [139]:
tipos_de_datos = Delay_Table.dtypes
print('Los tipos de datos de cada columna son:\n') 
tipos_de_datos

Los tipos de datos de cada columna son:



Unnamed: 0             int64
Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object

### Para las columnas que me plantean dudas:
- Visualizar los datos de las columnas y/o analizarlos con cálculo simple, 
- Calcular el tipo de dato de las columnas, sus medias, desviación típíca, y ver cómo se repiten los valores:
- saber si hay o no datos vacíos 
- y tomar decisioneso sobre ellos:

In [7]:
print('COLUMNA_1: Unnamed: 0')
print ('Los valores que aparecen en esta columna son:')
Delay_Table['Unnamed: 0']

# Tienen pinta de ser datos únicos: los Unique key de la tabla. No se puede prescindir de esta columna

COLUMNA_1: Unnamed: 0
Los valores que aparecen en esta columna son:


0                0
1                1
2                2
3                4
4                5
            ...   
1936753    7009710
1936754    7009717
1936755    7009718
1936756    7009726
1936757    7009727
Name: Unnamed: 0, Length: 1936758, dtype: int64

In [8]:
# Calcular el tipo de dato de las columnas, sus medias, desviación típíca, y ver cómo se repiten los valores:

print('COLUMNA_2: Year')
columna= Delay_Table['Year']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Parece que solo hay un valor, el 2008. Es una columna de la que se puede prescindir.

COLUMNA_2: Year
Tipo de dato: int64
Media: 2008.0
Desviación típica: 0.0

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


2008    1936758
Name: Year, dtype: int64

In [9]:

print('COLUMNA_3: Month')
columna= Delay_Table['Month']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# En este caso aparecen representados los 12 meses del año

COLUMNA_3: Month
Tipo de dato: int64
Media: 6.11110629206127
Desviación típica: 3.482546393665923

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


12    203385
6     200914
3     200842
2     189534
1     183527
7     182945
8     162648
4     155264
5     153072
11    105563
10    103525
9      95539
Name: Month, dtype: int64

In [10]:

print('COLUMNA_5: DayOfWeek')
columna= Delay_Table['DayOfWeek']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Aparecen los 7 días de la semana codificados del 1 al 7

COLUMNA_5: DayOfWeek
Tipo de dato: int64
Media: 3.9848272215733718
Desviación típica: 1.9959662750518143

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


5    323259
1    290933
4    289451
7    286111
3    262805
2    260943
6    223256
Name: DayOfWeek, dtype: int64

In [11]:
# Las columas 6 a 10: DepTime / CRSDepTime / ArrTime / CRSArrTime indican: 
# los tiempos de salida y llegada reales
# los tiempos estimados de salida y llegada.
# Son columnas importantes para mantener

In [12]:

print('COLUMNA_10: UniqueCarrier')
columna= Delay_Table['UniqueCarrier']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Son las compañías aéreas. Es una columna a mantener.

COLUMNA_10: UniqueCarrier
Tipo de dato: object

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


WN    377602
AA    191865
MQ    141920
UA    141426
OO    132433
DL    114238
XE    103663
CO    100195
US     98425
EV     81877
NW     79108
FL     71284
YV     67063
B6     55315
OH     52657
9E     51885
AS     39293
F9     28269
HA      7490
AQ       750
Name: UniqueCarrier, dtype: int64

In [13]:
print('COLUMNA_11: FlightNum')
columna= Delay_Table['FlightNum']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Son los número de vuelos. En total hay 7499 números de vuelos diferentes. Es una columna a mantener

COLUMNA_11: FlightNum
Tipo de dato: int64

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


16      1586
50      1550
40      1507
15      1471
44      1447
        ... 
7681       1
7032       1
6893       1
6914       1
5272       1
Name: FlightNum, Length: 7499, dtype: int64

In [14]:

print('COLUMNA_12: TailNum')
columna= Delay_Table['TailNum']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# TailNum es le número de cola de una aeronave. Es un identificador único asignado a cada avión.
# Hay que mantener la columna, puede aportar mucha información relevante.
# En total son 5366 aviones diferentes

COLUMNA_12: TailNum
Tipo de dato: object

El total de NaN en esta columna son: 5

Los valores que aparecen en esta columna son:


N325SW    965
N676SW    951
N658SW    945
N313SW    937
N308SA    936
         ... 
9189E       1
N853NW      1
N856NW      1
N76010      1
N78008      1
Name: TailNum, Length: 5366, dtype: int64

In [15]:

print('COLUMNA_13: ActualElapsedTime')
columna= Delay_Table['ActualElapsedTime']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Parece que indica la duración total del vuelo, por lo que debe ser una columna relevante.

COLUMNA_13: ActualElapsedTime
Tipo de dato: float64
Media: 133.30586334268665
Desviación típica: 72.06006897518155

El total de NaN en esta columna son: 8387

Los valores que aparecen en esta columna son:


80.0     16871
75.0     16788
70.0     16032
78.0     15857
77.0     15806
         ...  
686.0        1
687.0        1
727.0        1
682.0        1
790.0        1
Name: ActualElapsedTime, Length: 673, dtype: int64

In [16]:

print('COLUMNA_14: CRSElapsedTime')
columna= Delay_Table['CRSElapsedTime']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Parece que indica los tiempos estimados para cada vuelo. 
# También parece una columna relevante para hacer cálculos sobre la desviación final de cada vuelo.

COLUMNA_14: CRSElapsedTime
Tipo de dato: float64
Media: 134.3027440409799
Desviación típica: 71.34143888170351

El total de NaN en esta columna son: 198

Los valores que aparecen en esta columna son:


75.0     50736
70.0     49950
65.0     49738
85.0     48934
80.0     47213
         ...  
453.0        1
449.0        1
475.0        1
14.0         1
602.0        1
Name: CRSElapsedTime, Length: 515, dtype: int64

In [17]:

print('COLUMNA_15: AirTime')
columna= Delay_Table['AirTime']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# AirTime se refiere al tiempo real que pasa en el aire desde el despegue hasta el aterrizaje. 
# Esta variable excluye el tiempo en tierra, como rodaje en pista o espera en pista antes del despegue.
# También parece una columna relevante para hacer cálculos.

COLUMNA_15: AirTime
Tipo de dato: float64
Media: 108.27714739539228
Desviación típica: 68.64261013867596

El total de NaN en esta columna son: 8387

Los valores que aparecen en esta columna son:


45.0     18374
46.0     18084
52.0     18052
44.0     18019
53.0     17995
         ...  
632.0        1
639.0        1
652.0        1
655.0        1
585.0        1
Name: AirTime, Length: 650, dtype: int64

In [18]:

print('COLUMNA_16: ArrDelay')
columna= Delay_Table['ArrDelay']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# ArrDelay indica el retraso en minutos que experimentó el vuelo al llegar a su destino.
# Es fruto del cálculo entre ArrTime y CRSArrTime
# Es una columna prescindible porque si se necesita el dato se puede volver a calcular
# Lo mismo pasa con DepDelay, la columna 17, que también se puede prescindir.


COLUMNA_16: ArrDelay
Tipo de dato: float64
Media: 42.19988477321014
Desviación típica: 56.78471513732212

El total de NaN en esta columna son: 8387

Los valores que aparecen en esta columna son:


10.0      36291
8.0       35902
7.0       35869
9.0       35782
6.0       35522
          ...  
969.0         1
1707.0        1
1018.0        1
734.0         1
991.0         1
Name: ArrDelay, Length: 1128, dtype: int64

In [19]:
       
print('COLUMNA_18: Origin')
columna= Delay_Table['Origin']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Ciudades de origen. Es oportuno mantenerla

COLUMNA_18: Origin
Tipo de dato: object

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


ATL    131613
ORD    125979
DFW     95414
DEN     74323
LAX     58772
        ...  
BJI         4
PIR         3
PUB         2
INL         1
TUP         1
Name: Origin, Length: 303, dtype: int64

In [20]:

print('COLUMNA_19: Dest')
columna= Delay_Table['Dest']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Ciudades de destino. Es oportuno mantenerla

COLUMNA_19: Dest
Tipo de dato: object

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


ORD    108984
ATL    106898
DFW     70657
DEN     63003
LAX     59969
        ...  
INL         9
PIR         3
CYS         1
TUP         1
OGD         1
Name: Dest, Length: 304, dtype: int64

In [21]:

print('COLUMNA_20: Distance')
columna= Delay_Table['Distance']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# La distancia que separa origen y destino. Interesa mantenerla

COLUMNA_20: Distance
Tipo de dato: int64
Media: 765.6861590348407
Desviación típica: 574.4796530720855

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


337    13182
733     8720
761     8700
296     8652
403     8623
       ...  
75         1
905        1
746        1
111        1
99         1
Name: Distance, Length: 1419, dtype: int64

In [22]:
           
print('COLUMNA_21: TaxiIn')
columna= Delay_Table['TaxiIn']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()


# El tiempo que el avión rueda por la pista a la llegada

COLUMNA_21: TaxiIn
Tipo de dato: float64
Media: 6.812975216205236
Desviación típica: 5.273595064525684

El total de NaN en esta columna son: 7110

Los valores que aparecen en esta columna son:


4.0      323427
5.0      314063
6.0      239676
3.0      234686
7.0      177889
          ...  
165.0         1
174.0         1
156.0         1
149.0         1
173.0         1
Name: TaxiIn, Length: 180, dtype: int64

In [23]:
print('COLUMNA_22: TaxiOut')
columna= Delay_Table['TaxiOut']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# El tiempo que el avión rueda por la pista a la salida

COLUMNA_22: TaxiOut
Tipo de dato: float64
Media: 18.232202811233574
Desviación típica: 14.338534198201843

El total de NaN en esta columna son: 455

Los valores que aparecen en esta columna son:


11.0     129633
10.0     129556
12.0     126355
13.0     119526
9.0      118894
          ...  
422.0         1
291.0         1
254.0         1
309.0         1
370.0         1
Name: TaxiOut, Length: 332, dtype: int64

In [24]:

print('COLUMNA_23: Cancelled')
columna= Delay_Table['Cancelled']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Es un código binario (0,1) que indica si el vuelo fue cancelado o no
# El porcentaje de cancelados es muy bajo. No merece la pena mantener la columna.

COLUMNA_23: Cancelled
Tipo de dato: int64
Media: 0.00032683484462178546
Desviación típica: 0.018075624257674706

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


0    1936125
1        633
Name: Cancelled, dtype: int64

In [25]:

print('COLUMNA_24: CancellationCode')
columna= Delay_Table['CancellationCode']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Indica el motivo de la cancelación. Hay 3 motivos diferentes y los N son los no cancelados.
# Solo es interesante mantener para sacar conclusiones sobre los motivos de la cancelación, 
# También es un argumento que me confirma para eliminar la columna anterior, 
# puesto que toda la información sobre la cancelación está recogida en esta columna.

COLUMNA_24: CancellationCode
Tipo de dato: object

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


N    1936125
B        307
A        246
C         80
Name: CancellationCode, dtype: int64

In [26]:

print('COLUMNA_25: Diverted')
columna= Delay_Table['Diverted']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)
print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# Diverted indica mediante código binario (0,1) si el vuelo fue desviado o no.
# El número de vuelos desviados no es pequeño. Puede resultar interesante mantener la columna
# Aunque si se ha desviado el vuelo no tiene mucho sentido calcular retrasos con esos datos.
# Hay que valorar si se deja o se elimina. Quizás valga la pena elimiar las filas en las que hay vuelos desviados,
# porque sus valores pueden distorsionar significativamente el resultado final

COLUMNA_25: Diverted
Tipo de dato: int64
Media: 0.004003597764924684
Desviación típica: 0.06314721711010741

El total de NaN en esta columna son: 0

Los valores que aparecen en esta columna son:


0    1929004
1       7754
Name: Diverted, dtype: int64

In [27]:

print('COLUMNA_26: CarrierDelay')
columna= Delay_Table['CarrierDelay']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)

print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# CarrierDelay indica los minutos de retraso que son responsabilidad de la aerolínea.
# Existen muchos NaN, sobre los que hay que tomar una decisión.


COLUMNA_26: CarrierDelay
Tipo de dato: float64
Media: 19.179398920069772
Desviación típica: 43.54620724547479

El total de NaN en esta columna son: 689270

Los valores que aparecen en esta columna son:


0.0       591049
6.0        22799
7.0        21500
15.0       20361
8.0        19506
           ...  
928.0          1
940.0          1
684.0          1
1096.0         1
905.0          1
Name: CarrierDelay, Length: 983, dtype: int64

In [28]:

print('COLUMNA_27: WeatherDelay')
columna= Delay_Table['WeatherDelay']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)

print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# WeatherDelay indica los minutos de retraso que son responsabilidad de la climatología
# Existen muchos NaN, sobre los que hay que tomar una decisión.

COLUMNA_27: WeatherDelay
Tipo de dato: float64
Media: 3.7035706956700185
Desviación típica: 21.49290048410904

El total de NaN en esta columna son: 689270

Los valores que aparecen en esta columna son:


0.0      1148570
10.0        2868
6.0         2738
15.0        2709
7.0         2527
          ...   
435.0          1
516.0          1
553.0          1
523.0          1
396.0          1
Name: WeatherDelay, Length: 599, dtype: int64

In [29]:

print('COLUMNA_28: NASDelay')
columna= Delay_Table['NASDelay']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)

print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# NASDelay representa los retrasos por problemas en la National Airspace System
# Existen muchos NaN, sobre los que hay que tomar una decisión.

COLUMNA_28: NASDelay
Tipo de dato: float64
Media: 15.021635478657911
Desviación típica: 33.83305216365376

El total de NaN en esta columna son: 689270

Los valores que aparecen en esta columna son:


0.0      588995
1.0       34060
2.0       33177
3.0       31615
4.0       29825
          ...  
838.0         1
768.0         1
683.0         1
919.0         1
413.0         1
Name: NASDelay, Length: 574, dtype: int64

In [30]:

print('COLUMNA_29: SecurityDelay')
columna= Delay_Table['SecurityDelay']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)

print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# SecurityDelay representa los retrasos por problemas en la seguridad
# Existen muchos NaN, sobre los que hay que tomar una decisión.

COLUMNA_29: SecurityDelay
Tipo de dato: float64
Media: 0.09013713959573158
Desviación típica: 2.0227140475056054

El total de NaN en esta columna son: 689270

Los valores que aparecen en esta columna son:


0.0      1241483
7.0          350
6.0          327
8.0          286
9.0          278
          ...   
284.0          1
100.0          1
149.0          1
156.0          1
254.0          1
Name: SecurityDelay, Length: 156, dtype: int64

In [31]:

print('COLUMNA_30: LateAircraftDelay')
columna= Delay_Table['LateAircraftDelay']
tipo_de_dato = columna.dtype
print('Tipo de dato:', tipo_de_dato)

media = columna.mean()
print('Media:', media)

desviacion_tipica = columna.std()
print('Desviación típica:', desviacion_tipica)
print()
NaN_count = columna.isna().sum()
print ('El total de NaN en esta columna son:', NaN_count)

print()
print ('Los valores que aparecen en esta columna son:')
columna.value_counts()

# LateAircraftDelay representa los retrasos originados por las conexiones con otros vuelos
# Existen muchos NaN, sobre los que hay que tomar una decisión.

COLUMNA_30: LateAircraftDelay
Tipo de dato: float64
Media: 25.296466178432176
Desviación típica: 42.05486152095404

El total de NaN en esta columna son: 689270

Los valores que aparecen en esta columna son:


0.0       555906
15.0       18632
16.0       17315
17.0       16393
18.0       15550
           ...  
500.0          1
535.0          1
850.0          1
501.0          1
1000.0         1
Name: LateAircraftDelay, Length: 564, dtype: int64

### CONCLUSIONES SOBRE LAS COLUMNAS/VARIABLES:

La información que contiene cada una de las variables es: 

1.- __Unnamed: 0__: Puede ser un índice sin nombre: la __UNIQUE KEY__ de la tabla (Mantener).

2.- __Year__: Año del vuelo. Siempre es 2008 (Eliminar).

3.- __Month__: Mes del vuelo (Eliminar, no aporta información relevante para los cálculos).

4.- __DayofMonth__: Día del mes del vuelo (Eliminar, no aporta información relevante para los cálculos).

5.- __DayOfWeek__: Día de la semana del vuelo (Mantener, puede servír para ver qué día de la semana son mayores los retrasos).

6.- __DepTime__: Hora de salida real del vuelo (Eliminar. No voy a hacer cálculos con ellas porque hay otras columnas que contienen información más directa).

7.- __CRSDepTime__: Hora de salida programada del vuelo (Eliminar. No voy a hacer cálculos con ellas porque hay otras columnas que contienen información más directa).

8.- __ArrTime__: Hora de llegada real del vuelo (Eliminar. No voy a hacer cálculos con ellas porque hay otras columnas que contienen información más directa).

9.- __CRSArrTime__: Hora de llegada programada del vuelo (Eliminar. No voy a hacer cálculos con ellas porque hay otras columnas que contienen información más directa).

10.- __UniqueCarrier__: Código que identifica de manera única a una aerolínea (Mantener).

11.- __FlightNum__: Número de vuelo (Mantener).

12.- __TailNum__: Número de cola de la aeronave (Mantener).

13.- __ActualElapsedTime__: Tiempo real de vuelon(Mantener).

14.- __CRSElapsedTime__: Tiempo de vuelo programado (Mantener).

15.- __AirTime__: Tiempo de vuelo en el aire (Mantener).

16.- __ArrDelay__: Retraso en la llegada (Mantener).

17.- __DepDelay__: Retraso en la salida (Eliminar. No voy a hacer cálculos con ellas porque hay otras columnas que contienen información más directa).

18.- __Origin__: Aeropuerto de origen (Mantener).

19.- __Dest__: Aeropuerto de destino (Mantener).

20.- __Distance__: Distancia entre aeropuertos (Mantener).

21.- __TaxiIn__: Tiempo de rodaje en pista al aterrizar (Mantener).

22.- __TaxiOut__: Tiempo de rodaje en pista al despegar (Mantener).

23.- __Cancelled__: Indica si el vuelo fue cancelado (Eliminar, se puede obtener la info con CancellationCode).

24.- __CancellationCode__: Código que indica la razón de la cancelación (Mantener).

25.- __Diverted__: Indica si el vuelo fue desviado (Eliminar, porque distorsiona los resultados).

26.- __CarrierDelay__: Retraso atribuible a la aerolínea ((Mantener).

27.- __WeatherDelay__: Retraso atribuible a condiciones climáticas (Mantener).

28.- __NASDelay__: Retraso atribuible al sistema nacional de aviación (Mantener).

29.- __SecurityDelay__: Retraso atribuible a problemas de seguridad (Mantener).

30.- __LateAircraftDelay__: Retraso atribuible a una llegada de aeronave anterior (Mantener).

### CONCLUSIONES SOBRE CELDAS VACIAS: 

- En la COLUMNA_12: TailNum hay 5 valores vacíos sobre los que hay que tomar decisiones
- En la COLUMNA_13: ActualElapsedTime aparecen 8387 valores vacíos sobre los que hay que tomar decisión. Estos NaN pueden corresponderse con los vuelos cancelados, vuelos que en realidad no han volado, por lo que es mejor eliminar esas filas para que sus datos no distorsione los resultados finales. 
- En la COLUMNA_14: CRSElapsedTime aparecen 198 valores vacíos
- En la COLUMNA_15: AirTime aparecen 8387 valores vacíos. Idem que Columna_13.
- En la COLUMNA_16: ArrDelay aparecen 8387 valores vacíos. Idem que Columna_13 y 14.
- En la COLUMNA_21: TaxiIn aparecen 7110 valores vacíos
- En la COLUMNA_22: TaxiOut aparecen 455 valores vacíos
- EN las COLUMNAS 26 A 30 Los valores NaN son los mismos en todas las coumnas, por lo que interpreto que cuando hay un NaN, se repite para esas columnas en toda la fila.Por lo tanto, si queremos extraer conclusiones sobre las causas de los retrasos, esas filas son suscpetibles de ser eliminadas, porque no van a permitir realizar cálculos ni sacar conclusiones con ellas. Pero por otra parte, también podrían estar aportando información de si un vuelo ha llegado o no en hora. Por lo tanto parece que es más interesante mantener esas filas.  

### DECISIONES DE DEPURACIÓN DE LA TABLA:

1.- Decido eliminar las siguientes columnas: 
- COLUMNAS_2, 3 y 4: Year, Month y DayofMonth. El año es siempre el mismo y las otras dos variables no aportan información significativa para realizar cálculos. 
- COLUMNAS_6, 7, 8 y 9: DepTime, CRSDepTime, ArrTime y CRSArrTime: No voy a hacer cálculos con ellas porque hay otras columnas que contienen información más directa sobre los tiempos de retrasos.
- COLUMNA_17: DepDelay: Retraso en la salida. La elimino porque el retraso que realmente me interesa es el de la llegada. Que haya o no retraso en la salida es indiferente.
- COLUMNA_23: Cancelled (vuelos candelados) porque aporta poca inforMación y esa información también se puede sacar de la COLUMNA_24: CancellatioCode,
- COLUMNA_25: Diverted (vuelos deviados)porque el número de vuelos desviados son pocos y no aporta una información significativa. No obstabte, antes de borrar la columna elimino las filas de en las que aparecen vuelos desviados para que su valor de retrasos no me distorsione los cálculos posterirores. 

2.- Decido eliminar las siguientes filas: 
- Las filas con valor (1) en COLUMNA_25: Diverted
- Las filas con valor NaN en las COLUMNAS 12 a 16, 21 y 22

==> PARA SER EFICIENTE, PRIMERO ELIMINO LAS FILAS SELECCIONADAS Y LUEGO LAS COLUMNAS

3.- Finalmente reduzco la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres, tal y como me pide el enunciado del ejercicio.


In [33]:
# 1.- Elimar las filas con valor NaN en la COLUMNAS 12 a 16, 21 y 22
# Con dropna() para eliminar los valores NaN y con subset que indica las columnas en las que seleccionar los NaN

columnas_a_verificar = ['TailNum', 'ActualElapsedTime', 'CRSElapsedTime','AirTime','ArrDelay','TaxiIn','TaxiOut']
Delay_Table_1 = Delay_Table.dropna(subset=columnas_a_verificar)

num_filas_1 = Delay_Table_1.shape[0]
print ('la nueva tabla Delay_Table_1 ha quedado compuesta compuesta por:', num_filas_1,'filas')

la nueva tabla Delay_Table_1 ha quedado compuesta compuesta por: 1928368 filas


In [34]:
# 2.- Eliminar las filas con valor (1) en COLUMNA_25: Diverted
# Lo hago seleccionando en la tabla solo las filas en las las que el valor de Diverted==0 
Delay_Table_2 = Delay_Table_1[Delay_Table_1['Diverted'] == 0]

num_filas_2 = Delay_Table_2.shape[0]
print ('la nueva tabla Delay_Table_2 ha quedado compuesta compuesta por:', num_filas_2,'filas')

# Veo que la tabla no se ha reducido de tamaño, lo que quiere decir que todas las filas en las que 
# aparecían vuelos desviados, ya habían sido eliminadas porque tendrían un valor NaN en otra variable.

la nueva tabla Delay_Table_2 ha quedado compuesta compuesta por: 1928368 filas


In [121]:
# 3.- Eliminar las columnas que no me interesan: 
# Con drop(), e indicándole que es el eje 1, el de las columnas:

columnas_a_eliminar = ['Year','Month','DayofMonth', 'DepTime','CRSDepTime',
                       'ArrTime','CRSArrTime','DepDelay','Cancelled','Diverted']
Delay_Table_3 = Delay_Table_2.drop(columnas_a_eliminar, axis=1)

num_columnas_3 = Delay_Table_3.shape[1]
print ('la tabla Delay_Table_3 ha quedado constituida por un número final de:', num_columnas_3,'columnas')

la tabla Delay_Table_3 ha quedado constituida por un número final de: 20 columnas


In [122]:
# 4.- Reducir la dimensión de la tabla de manera aleatoria dejando solo 200.000 registros

filas_final = 200000  

# Obtengo una muestra aleatoria de n filas con sample()
filas_a_mantener = Delay_Table_3.sample(filas_final) 

# Elimino las filas que no están en la muestra aleatoria con drop () y difrerence()
# Con la extensión .index me aseguro que que la tabla final solo contenga las filas que están en la muestra aleatoria
Delay_Final = Delay_Table_3.drop(Delay_Table_3.index.difference(filas_a_mantener.index))

num_filas_Final = Delay_Final.shape[0]
print ('la nueva tabla Delay_Final ha quedado compuesta compuesta por:', num_filas_Final,'filas')

la nueva tabla Delay_Final ha quedado compuesta compuesta por: 200000 filas


In [141]:
# Visualizo una muestra de la tabla de los datos de principio, de fin y del centro
print(Delay_Final.head())

print(Delay_Final.iloc[100000:100010])

print(Delay_Final.tail())

    Unnamed: 0  DayOfWeek UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
4            5          4            WN        378  N726SW              101.0   
7           11          4            WN       1333  N334SW              121.0   
39          58          4            WN       2195  N479WN               86.0   
50          81          4            WN        890  N618WN               74.0   
61          96          4            WN       3940  N434WN              196.0   

    CRSElapsedTime  AirTime  ArrDelay Origin  ... TaxiIn  TaxiOut  \
4            115.0     87.0      11.0    IND  ...    4.0     10.0   
7            135.0    107.0      80.0    IND  ...    6.0      8.0   
39            90.0     72.0      26.0    JAX  ...    5.0      9.0   
50            80.0     60.0      77.0    LAS  ...    4.0     10.0   
61           205.0    177.0      60.0    LAS  ...    5.0     14.0   

    CancellationCode  CarrierDelay WeatherDelay  NASDelay  SecurityDelay  \
4                  N  

## 2.- Exercici 2
Fes un informe complet del dataset:

- Resumeix estadísticament el dataset i les columnes d’interès. 
- Fes una anàlisi estadístic del que consideris rellevant.
- Troba quantes dades faltants hi ha per columna.
- Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
- Fes una taula de les aerolínies amb més endarreriments acumulats.
- Quins són els vols més llargs? 
- I els més endarrerits? 
- Busca les rutes més llargues i les que acumulen més retards.
- Aporta allò que consideris rellevant.



In [125]:
# RESUMEN ESTADÍSTICO de las columnas de interés, con describe():

columnas_de_interes = ['ActualElapsedTime','CRSElapsedTime', 'AirTime','ArrDelay','Distance',
                       'TaxiIn','TaxiOut','CarrierDelay','WeatherDelay','NASDelay','SecurityDelay',
                       'LateAircraftDelay']
Resumen_estadistico = Delay_Final[columnas_de_interes].describe()

Resumen_estadistico


ActualElapsedTime  CRSElapsedTime        AirTime       ArrDelay  \
count      200000.000000   200000.000000  200000.000000  200000.000000   
mean          133.410680      134.320595     108.357600      42.262975   
std            72.137107       71.291116      68.650554      56.942219   
min            16.000000       15.000000       0.000000     -55.000000   
25%            80.000000       82.000000      58.000000       9.000000   
50%           116.000000      117.000000      90.000000      24.000000   
75%           165.000000      165.000000     137.000000      56.000000   
max           693.000000      660.000000     645.000000    1350.000000   

            Distance         TaxiIn        TaxiOut   CarrierDelay  \
count  200000.000000  200000.000000  200000.000000  129402.000000   
mean      766.164345       6.807290      18.245790      19.296526   
std       574.477202       5.293926      14.371294      43.560193   
min        11.000000       0.000000       0.000000       0.000000   
25%       340.000000       4.000000      10.000000       0.000000   
50%       607.000000       6.000000      14.000000       2.000000   
75%       998.000000       8.000000      21.000000      21.000000   
max      4962.000000     176.000000     345.000000    1348.000000   

        WeatherDelay       NASDelay  SecurityDelay  LateAircraftDelay  
count  129402.000000  129402.000000  129402.000000      129402.000000  
mean        3.716851      14.990016       0.079241          25.295590  
std        21.132230      33.682856       1.675947          42.629913  
min         0.000000       0.000000       0.000000           0.000000  
25%         0.000000       0.000000       0.000000           0.000000  
50%         0.000000       2.000000       0.000000           8.000000  
75%         0.000000      14.000000       0.000000          33.000000  
max      1225.000000     865.000000     170.000000        1316.000000

### ANÁLISIS ESTADÍSTICO: 

- La media de ActualElapsedTime (Tiempo real de vuelo) es muy parecida a la media de CRSElapsedTime (Tiempo de vuelo programado), por lo que se puede concluir que, por lo general, los aviones cumplen con el horario programado, salvo excepciones.

- Los tiempos de rodaje en pista al despegar son muy superiores a los de aterrizaje, por lo que el rodaje en despegue puede ser una variable susceptible de mejora.

- La causa que más retrasos ocasiona es la atribuible a la conexión con un vuelo anterior. También son importantes, aunque en menor medida, las causas atribuibles a las copañías aéreas y al sistema nacional. Los efectos de la climatología en los retrasos son poco significativos y los atribuibles a la seguridad, prácticamente irrelevantes.


### DATOS QUE FALTAN EN LAS COLUMNAS

En la tabla __*Delay_Final*__ que yo he seleccionado, no existen datos faltantes en ninguna de las categorías relevantes, porque han sido depurados antes de construir la tabla final. 

Solamente en las 5 últimas columnas (que contienen los datos referenetes a las causas de los retrasos) se han mantenido los valores NaN. En estas columnas cuando aparece un valor NaN aparece en las 5 columnas, por lo que se estima que esas filas NaN pueden aportar información sobre los vuelos que no han sufrido retrasos.

El total de filas que contienen valores NaN en esas 5 columnas es: 200.000 - 129.202 = 70.798 filas


### CREA COLUMNAS NUEVAS 
- velocitat mitjana del vol, 
- si ha arribat tard o no...

In [126]:
# Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
# Voy a considerar retraso si el ArrDelay supera es mayor que cero.
Delay_Final['VelMedia'],Delay_Final['Retraso']=Delay_Final['Distance'] / Delay_Final['AirTime'], (Delay_Final['ArrDelay']>0)
Delay_Final

Unnamed: 0  DayOfWeek UniqueCarrier  FlightNum TailNum  \
4                 5          4            WN        378  N726SW   
7                11          4            WN       1333  N334SW   
39               58          4            WN       2195  N479WN   
50               81          4            WN        890  N618WN   
61               96          4            WN       3940  N434WN   
...             ...        ...           ...        ...     ...   
1936679     7009362          6            DL       1088  N136DL   
1936686     7009387          6            DL       1123  N3747D   
1936693     7009406          6            DL       1150  N6706Q   
1936708     7009454          6            DL       1209  N903DA   
1936709     7009457          6            DL       1212  N703TW   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay Origin  ...  \
4                    101.0           115.0     87.0      11.0    IND  ...   
7                    121.0           135.0    107.0      80.0    IND  ...   
39                    86.0            90.0     72.0      26.0    JAX  ...   
50                    74.0            80.0     60.0      77.0    LAS  ...   
61                   196.0           205.0    177.0      60.0    LAS  ...   
...                    ...             ...      ...       ...    ...  ...   
1936679              250.0           228.0    183.0      46.0    LAS  ...   
1936686              208.0           222.0    173.0     374.0    LAS  ...   
1936693              105.0           116.0     81.0     103.0    SAN  ...   
1936708              168.0           124.0     92.0     140.0    SLC  ...   
1936709              322.0           312.0    284.0      23.0    OGG  ...   

        TaxiIn  TaxiOut  CancellationCode  CarrierDelay WeatherDelay  \
4          4.0     10.0                 N           NaN          NaN   
7          6.0      8.0                 N           8.0          0.0   
39         5.0      9.0                 N           0.0          0.0   
50         4.0     10.0                 N           7.0          0.0   
61         5.0     14.0                 N           0.0          0.0   
...        ...      ...               ...           ...          ...   
1936679   18.0     49.0                 N           0.0          0.0   
1936686    5.0     30.0                 N         374.0          0.0   
1936693    7.0     17.0                 N           0.0          0.0   
1936708    6.0     70.0                 N          96.0          0.0   
1936709   14.0     24.0                 N          13.0          0.0   

         NASDelay  SecurityDelay  LateAircraftDelay  VelMedia  Retraso  
4             NaN            NaN                NaN  7.908046     True  
7             0.0            0.0               72.0  7.738318     True  
39            0.0           10.0               16.0  6.722222     True  
50            0.0            0.0               70.0  8.116667     True  
61           22.0            0.0               38.0  8.971751     True  
...           ...            ...                ...       ...      ...  
1936679      22.0            0.0               24.0  9.546448     True  
1936686       0.0            0.0                0.0  9.699422     True  
1936693       0.0            0.0              103.0  7.728395     True  
1936708      44.0            0.0                0.0  6.847826     True  
1936709      10.0            0.0                0.0  8.753521     True  

[200000 rows x 22 columns]

### TABLA CON LAS AEROLÍNEAS CON MÁS RETRASOS ACUMULADOS.

In [144]:
# Agrupo por aerolínea (Columna UniqueCarrier) con .groupby() 
# y sumo los retrasos acumulados de la columna ArrDelay con .sum()
tabla_retrasos_acumulados = Delay_Final.groupby('UniqueCarrier')['ArrDelay'].sum()

# Los orordeno de mayor a menor con .sort_values() y con el argumento de ascending=False
tabla_retrasos_acumulados = tabla_retrasos_acumulados.sort_values(ascending=False)

print('Los retrasos que acumula cada una de las aerolíneas es:\n', tabla_retrasos_acumulados)

Los retrasos que acumula cada una de las aerolíneas es:
 UniqueCarrier
WN    1151875.0
AA     947837.0
UA     701966.0
MQ     651948.0
OO     627971.0
XE     541979.0
DL     474985.0
CO     420756.0
EV     406748.0
YV     375588.0
US     370380.0
NW     354562.0
FL     322891.0
B6     311382.0
OH     274833.0
9E     256288.0
AS     152234.0
F9      78599.0
HA      27591.0
AQ       2182.0
Name: ArrDelay, dtype: float64


### ¿CUÁLES HAN SIDO LOS VUELOS MÁS LARGOS (tiempo de vuelo)?

In [138]:
# Ordeno los por la columna 'ActualElapsedTime' (tiempo total del vuelo) de forma descendente con .sort_values() 
# y el argumento ascending=False
vuelos_mas_largos = Delay_Final.sort_values(by='ActualElapsedTime', ascending=False)

# Le pido que me imprima los 20 primeros vuelos.
n = 20
vuelos_mas_largos[['FlightNum', 'ActualElapsedTime','Origin', 'Dest', 'Distance']].head(n)

FlightNum  ActualElapsedTime Origin Dest  Distance
1065857         15              693.0    EWR  HNL      4962
556378          15              686.0    EWR  HNL      4962
1417613         15              683.0    EWR  HNL      4962
253995          29              679.0    JFK  LAX      2475
1080876        137              675.0    JFK  LAX      2475
718115          15              669.0    EWR  HNL      4962
360599          15              662.0    EWR  HNL      4962
366335          15              659.0    EWR  HNL      4962
1730191         15              657.0    EWR  HNL      4962
1066234         15              657.0    EWR  HNL      4962
1927700         15              653.0    EWR  HNL      4962
1255336         15              650.0    EWR  HNL      4962
1928363         15              649.0    EWR  HNL      4962
176213          15              649.0    EWR  HNL      4962
1619333         15              647.0    EWR  HNL      4962
174075          15              645.0    EWR  HNL      4962
1724446         15              642.0    EWR  HNL      4962
1930579         15              637.0    EWR  HNL      4962
872325          15              637.0    EWR  HNL      4962
1068615         15              635.0    EWR  HNL      4962

### ¿ Y LOS 10 VUELOS MÁS CORTOS?

In [146]:
vuelos_mas_cortos = Delay_Final.sort_values(by='ActualElapsedTime', ascending=False)

# Le pido que me imprima los últimos vuelos.
vuelos_mas_cortos[['FlightNum', 'ActualElapsedTime','Origin', 'Dest', 'Distance']].tail(10)

FlightNum  ActualElapsedTime Origin Dest  Distance
1060411         77               21.0    GST  JNU        41
1513140         64               21.0    PSG  WRG        31
229755        1010               21.0    KOA  HNL       163
864463          64               19.0    PSG  WRG        31
1513314         65               19.0    WRG  PSG        31
1406161         64               19.0    PSG  WRG        31
1240801         64               18.0    PSG  WRG        31
1910826         64               18.0    PSG  WRG        31
1240671         65               17.0    WRG  PSG        31
1058396         65               16.0    WRG  PSG        31

### ¿CUÁLES SON LOS VUELOS QUE MÁS RESTRASO TIENEN?

In [148]:
# Ordeno por la columna 'ArrDelay' de forma descendente
vuelos_mas_retrasados = Delay_Final.sort_values(by='ArrDelay', ascending=False)

# Imprimo los 10 primeros vuelos
n = 10
vuelos_mas_retrasados[['FlightNum','UniqueCarrier', 'Origin', 'Dest', 'ArrDelay']].head(n)

FlightNum UniqueCarrier Origin Dest  ArrDelay
318717         218            NW    PDX  MSP    1350.0
1027010        776            NW    LAS  MSP    1344.0
1222506        284            AA    HNL  LAX    1325.0
1464385         68            UA    LIH  LAX    1322.0
349850        2233            AA    RSW  STL    1312.0
1025817        610            NW    GEG  MSP    1289.0
836938         491            NW    TPA  DTW    1263.0
1221849        123            AA    DFW  HNL    1263.0
682367         590            NW    PDX  MSP    1249.0
1906664       1790            AA    LAS  STL    1225.0

### CUÁLES SON LAS AERONAVES QUE ACUMULAN MÁS RETRASOS?

In [150]:
# Agrupo por 'TailNum' y 'UniqueCarrier' (para que me aparezca también la aerolínea) 
# suma los retrasos acumulados de la columna 'ArrDelay'
aeronaves_con_retraso = Delay_Final.groupby(['TailNum', 'UniqueCarrier'])['ArrDelay'].sum()

# Ordeno de mayor a menor
aeronaves_con_retraso = aeronaves_con_retraso.sort_values(ascending=False)

# Imprimo los 10 primeros valores de la tabla
aeronaves_con_retraso.head(10)

TailNum  UniqueCarrier
N77331   YV               7040.0
N651ML   YV               6698.0
N521LR   YV               6233.0
N154SF   YV               6110.0
N75992   YV               6060.0
N87353   YV               5891.0
N75994   YV               5740.0
N27173   YV               5495.0
N17175   YV               5431.0
N570ML   YV               5390.0
Name: ArrDelay, dtype: float64

### ¿Y LAS AERONAVES QUE MENOS RETRASOS TIENEN?

In [151]:
aeronaves_con_retraso.tail(10)

TailNum  UniqueCarrier
N37018   CO               -4.0
N798UA   UA               -6.0
N199UA   UA               -7.0
N7AGAA   AA               -7.0
N860DA   DL               -8.0
N218UA   UA               -9.0
N199DN   DL              -13.0
N176DN   DL              -20.0
N216UA   UA              -22.0
N738AL   AQ              -34.0
Name: ArrDelay, dtype: float64

### BUSCAR LAS RUTAS MÁS LARGAS Y LAS QUE ACUMULAN MÁS RETRASOS.

In [136]:
# Agrupo por 'Distance', 'Origin' y 'Dest' con .groupby()
# Calculo la suma de los retrasos acumulados con la columna 'ArrDelay' y la agrecon con .agg()
rutas_mas_largas_y_retrasos = Delay_Final.groupby(['Distance', 'Origin', 'Dest']).agg({'ArrDelay': 'sum'})

# Para saber las rutas más largas y los retrasos que acumulan, ordeno de mayor a menor por distancia
rutas_mas_largas_y_retrasos = rutas_mas_largas_y_retrasos.sort_values(by='Distance', ascending=False)

# Imprime la tabla resultante
print('Las rutas más largas y sus retrasos acumulados son:\n',rutas_mas_largas_retrasos.head(10))

Las rutas más largas y sus retrasos acumulados son:
                       ArrDelay
Distance Origin Dest          
4962     HNL    EWR      214.0
         EWR    HNL      330.0
4502     HNL    ATL      391.0
         ATL    HNL     1318.0
4243     ORD    HNL     1390.0
         HNL    ORD      336.0
4213     KOA    ORD      380.0
4184     ORD    OGG      480.0
3972     MSP    HNL      198.0
         HNL    MSP     1109.0


In [137]:
# Para saber las rutas que acumulan más retrasos y sus distancias , ordeno de mayor a menor por {'ArrDelay': 'sum'}
rutas_mas_retrasadas = rutas_mas_largas_y_retrasos.sort_values(by='ArrDelay', ascending=False)

# Imprime la tabla resultante
print('Las rutas que acumulan más retrasos son:\n', rutas_mas_retrasadas.head(10))

Las rutas que acumulan más retrasos son:
                       ArrDelay
Distance Origin Dest          
733      ORD    LGA    26744.0
337      LAX    SFO    24630.0
733      LGA    ORD    22188.0
745      ATL    EWR    21754.0
761      ATL    LGA    20895.0
719      ORD    EWR    20165.0
337      SFO    LAX    19230.0
719      EWR    ORD    18495.0
802      DFW    ORD    18143.0
745      EWR    ATL    17414.0


## 3.- Exercici 3
Exporta el dataset net i amb les noves columnes a Excel.

1.- Compruebo mi ruta con __pwd__.

2.- Instalo el módulo __openpyxl__ porque me daba error.

3.- Creo la ruta y lo exporto con la función __.to_excel()__, indicándole index=False para que no me exporte el índice.

In [112]:
pwd

'/Users/Esteban/Jupyter Notebook'

In [117]:
pip install openpyxl

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 250.0/250.0 kB 3.7 MB/s eta 0:00:00a 0:00:01

[notice] A new release of pip available: 22.2.2 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [152]:
ruta='/Users/Esteban/Jupyter Notebook/Delay_Final.xlsx'

Delay_Final.to_excel(ruta, index=False)